build/sdk_update.go

```diff
@@ -0,0 +1,316 @@
+package build
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"unicode"
+
+	"gopkg.in/bblfsh/sdk.v2/assets/skeleton"
+	"gopkg.in/bblfsh/sdk.v2/driver/manifest"
+)
+
+const (
+	tplExt      = ".tpl"
+	manifestTpl = manifest.Filename + tplExt
+)
+
+var overwriteManagedFiles = os.Getenv("BABELFISH_OVERWRITE_MANAGED") == "true"
+
+// managedFiles are files that always are overwritten
+var managedFiles = map[string]bool{
+	".travis.yml":             true,
+	"Makefile":                true,
+	"update.go":               true,
+	"build.go":                true,
+	"test.go":                 true,
+	"README.md" + tplExt:      true,
+	"LICENSE":                 true,
+	"driver/main.go" + tplExt: true,
+	"driver/sdk_test.go":      true,
+	"driver/normalizer/transforms.go" + tplExt: true,
+}
+
+type updater struct {
+	root    string
+	opt     UpdateOptions
+	context map[string]interface{}
+	changes int
+}
+
+// PrintfFunc is a logging function type similar to log.Printf.
+type PrintfFunc func(format string, args ...interface{}) (int, error)
+
+// Silencef is a logging function that does nothing.
+func Silencef(format string, args ...interface{}) (int, error) {
+	return 0, nil
+}
+
+// UpdateOptions is a set of options available for
+type UpdateOptions struct {
+	DryRun bool
+
+	Debugf   PrintfFunc
+	Noticef  PrintfFunc
+	Warningf PrintfFunc
+}
+
+// ErrChangesRequired is returned by SDKUpdate in DryRun mode when changes are required.
+var ErrChangesRequired = errors.New("changes are required")
+
+// GenerateManifest writes the manifest file to a root driver directory.
+func GenerateManifest(root string, language string) error {
+	tpl := string(skeleton.MustAsset(manifestTpl))
+
+	t, err := template.New(manifestTpl).Funcs(funcs).Parse(tpl)
+	if err != nil {
+		return err
+	}
+
+	name := fixGitFolder(manifestTpl)
+	file := filepath.Join(root, manifest.Filename)
+
+	if _, err = os.Stat(file); err == nil {
+		return errors.New("trying to overwrite the manifest")
+	}
+	info, _ := skeleton.AssetInfo(name)
+
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_TRUNC, info.Mode())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
```

Comment:
It is safe for files. In other cases, I will have to copy-paste the `f.Close` to each error return by hand.

Reply:
It happens to work for an `*os.File`, but isn't guaranteed by the interface _or_ promised by the implementation. Since there are only two cases here, what I would recommend is:

```go
err := t.Execute(f, map[string]string{"Language": language})
cerr := f.Close()
if err != nil {
   return err
}
return cerr
```

That doesn't always work nicely if there are multiple routes, but in this case the terminus is close. More generally what I would advise in cases where there isn't an easy resolution is:

```go
func F(name string) (err error) {
    f, err := os.OpenFile(…)
    …
    defer func() {
       if cerr := f.Close(); cerr != nil && err == nil {
         err = cerr
       }
    }()
}
```

which is admittedly not pretty, but avoids the unsupported requirement. (Here, clearly, we don't need that).

Reply:
Sure, it isn't exactly mentioned in interface docs, but if you check, almost every type actually can be closed twice, not only `os.File`. Same is true for Go libraries in the wild. It's a very common pattern.

Anyway, copy-pasted it for each return. Expect more FD leak bugs in the future :confused: